offchain-modules/packages/x/src/handlers/eth.ts

```diff
@@ -1,63 +1,191 @@
+      return;
+    }
+    const confirmNumber = ForceBridgeCore.config.eth.confirmNumber;
+    const getLogBatchSize = 100;
+    for (;;) {
+      const currentBlockHeight = await this.ethChain.getCurrentBlockNumber();
+      if (currentBlockHeight === this.lastHandledBlockHeight) {
+        break;
+      }
+
+      for (;;) {
+        const endBlockNumber =
+          this.lastHandledBlockHeight + getLogBatchSize > currentBlockHeight
+            ? currentBlockHeight
+            : this.lastHandledBlockHeight + getLogBatchSize;
+
+        logger.info(`EthHandler init getLogs from:${this.lastHandledBlockHeight} to:${endBlockNumber}`);
+
+        await this.ethChain.getLogs(this.lastHandledBlockHeight, endBlockNumber - 1, async (log, parsedLog) => {
+          await this.onLogs(log, parsedLog);
+        });
+        const endBlock = await this.ethChain.getBlock(endBlockNumber);
+        await this.setLastHandledBlock(endBlockNumber, endBlock.hash);
+        await this.confirmEthLocks(endBlockNumber, confirmNumber);
+        if (endBlockNumber == currentBlockHeight) {
+          break;
         }
-        await this.db.createEthLock([
-          {
-            txHash: log.transactionHash,
-            amount: amount,
-            token: parsedLog.args.token,
-            recipient: uint8ArrayToString(fromHexString(parsedLog.args.recipientLockscript)),
-            sudtExtraData: parsedLog.args.sudtExtraData,
-            blockNumber: log.blockNumber,
-            blockHash: log.blockHash,
-            sender: parsedLog.args.sender,
-          },
-        ]);
-        logger.info(`EthHandler watchLockEvents save EthLock successful for eth tx ${log.transactionHash}.`);
-      } catch (e) {
-        logger.error(`EthHandler watchLockEvents error: ${e}`);
-        await asyncSleep(3000);
       }
+    }
+  }
+
+  async watchNewBlock() {
+    await this.ethChain.watchNewBlock(this.lastHandledBlockHeight, async (newBlock: ethers.providers.Block) => {
+      await this.onBlock(newBlock);
+    });
+  }
+
+  async onBlock(block: ethers.providers.Block) {
+    try {
+      const confirmNumber = ForceBridgeCore.config.eth.confirmNumber;
+      if (
+        confirmNumber !== 0 &&
+        this.lastHandledBlockHeight === block.number - 1 &&
+        this.lastHandledBlockHash !== '' &&
+        block.parentHash !== this.lastHandledBlockHash
+      ) {
+        logger.warn(
+          `EthHandler onBlock blockHeight:${block.number} parentHash:${block.parentHash} != lastHandledBlockHash:${
+            this.lastHandledBlockHash
+          } fork occur removeUnconfirmedLock events from:${block.number - confirmNumber}`,
+        );
+        await this.db.removeUnconfirmedLocks(block.number, confirmNumber);
+        await this.ethChain.getLogs(block.number - confirmNumber, block.number, async (log, parsedLog) => {
+          await this.onLogs(log, parsedLog);
+        });
+      }
+      await this.setLastHandledBlock(block.number, block.hash);
```

Comment:
`setLastHandledBlock` should be in the end of function.

Reply:
fixed